src/farbfeld.rs

```diff
@@ -133,6 +143,33 @@ impl<'a, R: 'a + Read> ImageDecoder<'a> for FarbfeldDecoder<R> {
     }
 }
 
+impl<'a, R: 'a + Read + Seek> ImageDecoderExt<'a> for FarbfeldDecoder<R> {
+    fn read_rect_with_progress<F: Fn(Progress)>(
+        &mut self,
+        x: u32,
+        y: u32,
+        width: u32,
+        height: u32,
+        buf: &mut [u8],
+        progress_callback: F,
+    ) -> ImageResult<()> {
+        println!("rrwp({}, {}, {}, {})", x, y, width, height);
+        // A "scanline" (defined as "shortest non-caching read" in the doc) is just one channel in this case
+
+        let start = self.reader.seek(SeekFrom::Current(0))?;
+        println!("start = {}", start);
+        image::load_rect(x, y, width, height, buf, progress_callback, self,
+                         |s, scanline| {
+                             println!("|s, scanline={}|", scanline);
+                             s.reader.seek(SeekFrom::Start(start + scanline * 2))?;
+                             Ok(())
+                         },
+                         |s, buf| {println!("|s, buf={:02x?}|", buf); s.reader.read(buf)})?;
```

Comment:
I think this should technically be `read_all`, though I admit it is unlikely that a 2 byte read wouldn't complete in a single call

Reply:
You likely mean `read_exact` and not `read_all`, a naming inconsistency with `Write`. Another problem of a simple `read` is that even very short read calls might return `Err(Interrupted)` for a variety of reasons and it's generally preferrable not to treat that error as fatal.

Reply:
Funnily enough: `load_rect()` takes an `F2: FnMut(&mut D, &mut [u8]) -> Result<usize, E>`, which leads to the abomination you see now:
```rust
s.reader.read_exact(buf).map(|()| buf.len())
```
HDR:
```rust
s.read_image_data(buf).map(|_| buf.len())
```
DXT (`read_scanline()` does a full read and returns `Ok(buf.len())`):
```rust
s.read_scanline(buf)
```
BMP
```rust
s.read_image_data(buf).map(|_| buf.len())
```

Worse still, `load_rect()` *doesn't check if the read anything*:
```rust
                if position >= start && end - position >= scanline_bytes {
                    read_scanline(decoder, &mut buf[(bytes_read as usize)..]
                                                   [..(scanline_bytes as usize)])?;
                    bytes_read += scanline_bytes;
                } else {
```
which is a Fun Thing to debug at best and plain dangerous at worst – the API strongly suggests to the user that it handles partial reads (in the style of `std::io::Read::read()`) but then just doesn't.

See top commit for what changes fixing the signature on that (thankfully internal) API brings about.